packages/git/src/node/dugite-git-watcher.ts

```diff
@@ -28,27 +30,53 @@ export class DugiteGitWatcherServer implements GitWatcherServer {
             watcher.dispose();
         }
         this.watchers.clear();
+        this.repositoryToWatchers.clear();
+        this.watcherToRepository.clear();
     }
 
     async watchGitChanges(repository: Repository): Promise<number> {
-        const watcher = this.manager.getWatcher(repository);
-        const disposable = new DisposableCollection();
-        disposable.push(watcher.onStatusChanged(e => {
-            if (this.client) {
-                this.client.onGitChanged(e);
-            }
-        }));
-        disposable.push(watcher.watch());
+        const repositoryWatcher = this.manager.getWatcher(repository);
+        const { localUri } = repository;
+        if (!this.watchers.has(localUri)) {
+            const disposable = new DisposableCollection();
+            disposable.push(repositoryWatcher.onStatusChanged(e => {
+                if (this.client) {
+                    this.client.onGitChanged(e);
+                }
+            }));
+            disposable.push(repositoryWatcher.watch());
```

Comment:
This method should be called for each call of `watchGitChanges`, otherwise, new clients won't get the initial state. It seems that now it is called only for once for each repository URL.

Reply:
> What was the root cause?

Well, we had two separate watchers after calling `watchChanges` twice.

You are right. The second `watchChanges` call will not notify once more. I need to work on this.

Reply:
> Well, we had two separate watchers after calling watchChanges twice.

Do you mean 2 instances of `GitRepositoryWatcher` for the same repository? `GitRepositoryManagerImpl.getWatcher` should not allow it.

Reply:
I see `MockRepositoryManager` is bogus.

Reply:
the issue is that there is only one watcher but for the same client multiple listeners are installed on it, I was able to fix it with the following `watchGitChanges`:
```ts
protected readonly watchers = new Map<number, Disposable>();
protected readonly subscriptions = new Map<string, DisposableCollection>();

constructor(
    @inject(GitRepositoryManager) protected readonly manager: GitRepositoryManager
) { }

dispose(): void {
    for (const watcher of this.watchers.values()) {
        watcher.dispose();
    }
    this.watchers.clear();
    this.subscriptions.clear();
}

async watchGitChanges(repository: Repository): Promise<number> {
    const watcher = this.manager.getWatcher(repository);

    const repositoryUri = repository.localUri;
    let subscriptions = this.subscriptions.get(repositoryUri);
    if (subscriptions === undefined) {
        const unsubscribe = watcher.onStatusChanged(e => {
            if (this.client) {
                this.client.onGitChanged(e);
            }
        });
        subscriptions = new DisposableCollection();
        subscriptions.onDispose(() => unsubscribe.dispose());
        this.subscriptions.set(repositoryUri, subscriptions);
    }

    const disposable = watcher.watch();
    subscriptions.push(disposable);
    const watcherId = this.watcherSequence++;
    this.watchers.set(watcherId, disposable);
    return watcherId;
}
```

It makes sure that only one event listener is installed by the same client for the same watcher and it is get removed when all watchers are stopped.

BTW What was the point to have mocks again? I see that tests can be rewritten without mocks at all.

Reply:
> I was able to fix it with the following

The way you have simplified the (flawed) logic I have written... Big thanks for that! I would like to add a test for the _"..otherwise; new clients won't get the initial state..."_ case too. Plus at least one for the listener removal.

> What was the point to have mocks again?

Tests started to fail randomly on the CI because the `after` phase was unable to clean the Git repository, a process held it. I wanted to decouple the Git logic from the listeners and process. That was all.

> deleted all of them to make sure that exactly the same code is used in tests

👍

Reply:
> Tests started to fail randomly on the CI because the after phase...

Was it because of `GitRepositoryManager.run`? Maybe we can mock only one method, like `manager.run = (_, op) => op()`.

Reply:
OK, let’s try that and see how it works on the CIs.